Python/ceval.c

```diff
@@ -1063,17 +1064,121 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+    int pc = next_uop - 1 - current_executor->trace;
+    _PyExecutorObject **pexecutor = current_executor->executors + pc;
+    if (*pexecutor != NULL) {
+#ifdef Py_DEBUG
+        PyCodeObject *code = _PyFrame_GetCode(frame);
+        DPRINTF(2, "Jumping to new executor for %s (%s:%d) at byte offset %d\n",
+                PyUnicode_AsUTF8(code->co_qualname),
+                PyUnicode_AsUTF8(code->co_filename),
+                code->co_firstlineno,
+                2 * (int)(frame->instr_ptr - _PyCode_CODE(_PyFrame_GetCode(frame))));
+#endif
+        Py_DECREF(current_executor);
+        current_executor = (_PyUOpExecutorObject *)*pexecutor;
+        Py_INCREF(current_executor);
+        goto enter_tier_two;
+    }
+
+    // Increment and check side exit counter.
+    // (Even though we only need it for certain opcodes.)
+    next_instr = frame->instr_ptr;
+    uint16_t *pcounter = current_executor->counters + pc;
+    *pcounter += 1 << OPTIMIZER_BITS_IN_COUNTER;
+    /* We are using unsigned values, but we really want signed values, so
+     * do the 2s complement comparison manually */
+    uint16_t ucounter = *pcounter + (1 << 15);
+    uint16_t threshold = tstate->interp->optimizer_resume_threshold + (1 << 15);
+    if (ucounter <= threshold)
+    {
+        Py_DECREF(current_executor);
+        goto resume_frame_using_stack_pointer;
+    }
+
+    // Decode instruction to look past EXTENDED_ARG.
+    opcode = next_instr[0].op.code;
+    if (opcode == EXTENDED_ARG) {
+        opcode = next_instr[1].op.code;
+    }
+
+    // For selected opcodes build a new executor and enter it now.
+    if (opcode == POP_JUMP_IF_FALSE ||
+        opcode == POP_JUMP_IF_TRUE ||
+        opcode == POP_JUMP_IF_NONE ||
+        opcode == POP_JUMP_IF_NOT_NONE)
+    {
+        DPRINTF(2, "--> %s @ %d in %p has %d side exits\n",
+                _PyUOpName(uopcode), pc, current_executor, (int)(*pcounter));
+        DPRINTF(2, "    T1: %s\n", _PyOpcode_OpName[opcode]);
+
+        int optimized = _PyOptimizer_Unanchored(frame, next_instr, pexecutor, stack_pointer);
+        if (optimized < 0) {
+            goto error_tier_two;
+        }
 
+        if (!optimized) {
+            DPRINTF(2, "--> Failed to optimize %s @ %d in %p\n",
+                    _PyUOpName(uopcode), pc, current_executor);
+        }
+        else {
+#ifdef Py_DEBUG
+            DPRINTF(1, "--> Optimized %s @ %d in %p\n",
+                    _PyUOpName(uopcode), pc, current_executor);
+            PyCodeObject *code = _PyFrame_GetCode(frame);
+            DPRINTF(2, "Jumping to fresh executor for %s (%s:%d) at byte offset %d\n",
+                    PyUnicode_AsUTF8(code->co_qualname),
+                    PyUnicode_AsUTF8(code->co_filename),
+                    code->co_firstlineno,
+                    2 * (int)(frame->instr_ptr - _PyCode_CODE(_PyFrame_GetCode(frame))));
+#endif
+            Py_DECREF(current_executor);
+            current_executor = (_PyUOpExecutorObject *)*pexecutor;
```

Comment:
This can access freed memory (*pexecutor points into an array owned by current_executor).

The fix would be a little more elaborate than for L1089, there's a use of `*pcounter` below as well, and another of `*pexecutor`.